subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/DefaultVisitedArtifactResultsTest.groovy

```diff
@@ -46,4 +46,71 @@ class DefaultVisitedArtifactResultsTest extends Specification {
         selected.getArtifactsWithId(1) == variant2Artifacts
     }
 
+    def "lenient selection includes selected variant of each node"() {
```

Comment:
Probably also worth adding tests that ensure strict selection fails with unavailable or broken artifact sets.

Reply:
Added.